src/Cryptography/LibSodium/FFI.hsc

```diff
@@ -0,0 +1,195 @@
+
+-- | @void sodium_init();@
+--
+-- <https://libsodium.gitbook.io/doc/usage>
+foreign import capi "sodium.h sodium_init"  c_sodium_init :: IO CInt
+
+-------------------------------------------------------------------------------
+-- Memory management
+-------------------------------------------------------------------------------
+
+-- | @void sodium_memzero(void * const pnt, const size_t len);@
+--
+-- <https://libsodium.gitbook.io/doc/memory_management#zeroing-memory>
+foreign import capi unsafe "sodium.h sodium_memzero" c_sodium_memzero :: Ptr a -> CSize -> IO ()
+
+-- | @void *sodium_malloc(size_t size);@
+--
+-- <https://libsodium.gitbook.io/doc/memory_management>
+foreign import capi unsafe "sodium.h sodium_malloc" c_sodium_malloc :: CSize -> IO (Ptr a)
+--
+-- | @void sodium_free(void *ptr);@
+--
+-- <https://libsodium.gitbook.io/doc/memory_management>
+foreign import capi unsafe "sodium.h sodium_free" c_sodium_free :: Ptr a -> IO ()
+
+-- | @void sodium_free(void *ptr);@
+--
+-- <https://libsodium.gitbook.io/doc/memory_management>
+foreign import capi unsafe "sodium.h &sodium_free" c_sodium_free_funptr :: FunPtr (Ptr a -> IO ())
+
+-------------------------------------------------------------------------------
+-- Hashing: SHA256
+-------------------------------------------------------------------------------
+
+-- | @int crypto_hash_sha256(unsigned char *out, const unsigned char *in, unsigned long long inlen);@
+--
+-- <https://libsodium.gitbook.io/doc/advanced/sha-2_hash_function>
+foreign import capi unsafe "sodium.h crypto_hash_sha256" c_crypto_hash_sha256 :: SizedPtr CRYPTO_SHA256_BYTES -> Ptr CUChar -> CULLong -> IO CInt
+
+-- | @int crypto_hash_sha256_init(crypto_hash_sha256_state *state);@
+foreign import capi unsafe "sodium.h crypto_hash_sha256_init" c_crypto_hash_sha256_init :: SizedPtr CRYPTO_SHA256_STATE_SIZE -> IO CInt
+
+-- | @int crypto_hash_sha256_update(crypto_hash_sha256_state *state, const unsigned char *in, unsigned long long inlen);@
+foreign import capi unsafe "sodium.h crypto_hash_sha256_update" c_crypto_hash_sha256_update :: SizedPtr CRYPTO_SHA256_STATE_SIZE -> Ptr CUChar -> CULLong -> IO CInt
+
+-- | @int crypto_hash_sha256_final(crypto_hash_sha256_state *state, unsigned char *out);@
+foreign import capi unsafe "sodium.h crypto_hash_sha256_final" c_crypto_hash_sha256_final :: SizedPtr CRYPTO_SHA256_STATE_SIZE -> SizedPtr CRYPTO_SHA256_BYTES -> IO CInt
+
+-------------------------------------------------------------------------------
+-- Hashing: Blake2b
+-------------------------------------------------------------------------------
+
+-- | @int crypto_generichash_blake2b(unsigned char *out, size_t outlen, const unsigned char *in, unsigned long long inlen, const unsigned char *key, size_t keylen);@
+--
+-- <https://libsodium.gitbook.io/doc/hashing/generic_hashing>
+foreign import capi unsafe "sodium.h crypto_generichash_blake2b" c_crypto_generichash_blake2b
```

Comment:
1) Out parameter, this is where the result of the hash will go
2) Length of the result
3) In parameter ,what you want to hash
4) Length of what you want to hash
5) Key 
6) Length of the key